lib/BuildSystem/BuildSystem.cpp

```diff
@@ -2444,25 +2444,27 @@ class StaleFileRemovalCommand : public Command {
 
       // If root paths are defined, stale file paths should be absolute.
       if (roots.size() > 0 && fileToDelete[0] != path_separator) {
-        bsci.getDelegate().error("", {}, (Twine("Stale file '") + fileToDelete + "' has a relative path. This is invalid in combination with the root path attribute."));
+        bsci.getDelegate().commandHadWarning(this, "Stale file '" + fileToDelete + "' has a relative path. This is invalid in combination with the root path attribute.");
         continue;
       }
 
       // Check if the file is located under one of the allowed root paths.
       for (auto root : roots) {
         auto res = std::mismatch(root.begin(), root.end(), fileToDelete.begin());
-        if (res.first == root.end() && ((*(res.first++) == '\0') || (*(res.first++) == path_separator))) {
+        if (res.first == root.end() && ((*(res.second++) == '\0') || (*(res.second++) == path_separator))) {
```

Comment:
It was discovered by a test which wasn't running before, because it relied on the command not failing immediately. This was the commented assertion with the `FIXME` in the `staleFileRemovalWithRoots` test.